### DISCUSSION TITLE
How to launch dnsmasq as dnsmasq user in a network namespace?

### DISCUSSION THREAD

**Discussion by crocket** [netblue30/firejail#4940] [Q&A] [answered]
Launching dnsmasq as dnsmasq user in a network namespace with firejail required

/etc/firejail/dnsmasq.local

```
# this includes `netns vpn`
include bridge-vpn.inc

# `netns vpn` requires
ignore nonewprivs
## Even `protocol unix,inet,inet6,netlink,packet,bluetooth`
## denies permission to create any listening socket
## with `netns vpn`
ignore protocol
ignore seccomp
```

/etc/firejail/globals.local

```
# private and private-cache require ${HOME} to not be /dev/null
ignore private
ignore private-cache
```

* Why does `netns vpn` require ignore=nonewprivs, ignore=protocol, and ignore=seccomp?
* private and private-cache shouldn't require ${HOME} to not be /dev/null.
* `include *.local` shouldn't require ${HOME} to not be /dev/null.

**Comment by SkewedZeppelin** [comment]
Just use systemd sandboxing, way more tightly integrated.

**Reply by crocket** [reply]
I don't use systemd.

**Comment by rusty-snake** [comment]
> Why does netns vpn require ignore=nonewprivs, ignore=protocol, and ignore=seccomp?

seccomp (and therefore protocol) implies nnp (if you are not root?).

Which error do you get if nnp is set?

**Reply by crocket** [reply]
```
dnsmasq: failed to create listening socket for port 53: Permission denied
```

**Reply by rusty-snake** [reply]
And if you don't use `netns vpn` you can run with nnp?

**Reply by crocket** [reply]
Yes.

**Reply by crocket** [reply]
With dnscrypt-proxy.profile, I needed

/etc/firejail/dnscrypt-proxy.local

```
# this includes `netns vpn`
include bridge-vpn.inc

# `netns vpn` requires
ignore nonewprivs
# These options require seccomp which implies nonewprivs
ignore protocol
ignore seccomp.drop
ignore memory-deny-write-execute
```

In a nutshell, it seems `netns xxx` cannot live with `nonewprivs` or anything that implies `nonewprivs` via seccomp.

**Reply by rusty-snake** [reply]
If you `ip netns exec setpriv --no-new-privs /usr/bin/dnsmasq` does it work? i.e. is this a firejail thing or a kernel restriction that saves you from doing bad things?

**Reply by crocket** [reply]
It turns out that I already executed `sysctl net.ipv4.ip_unprivileged_port_start=0` in the init network namespace.
I didn't do that in vpn network namespace. If `net.ipv4.ip_unprivileged_port_start=1024`, binding to port 53 requires some privilege.

What kind of privilege do applications need to bind to port 53 inside a firejail sandbox if `net.ipv4.ip_unprivileged_port_start=1024`?

**Reply by rusty-snake** [reply]
> **ip_unprivileged_port_start - INTEGER**
> 
> This is a per-namespace sysctl. It defines the first unprivileged port in the network namespace.  Privileged ports require root or CAP_NET_BIND_SERVICE in order to bind to them. To disable all privileged ports, set this to 0.  They must not overlap with the ip_local_port_range.
>
> Default: 1024

https://www.kernel.org/doc/html/latest/networking/ip-sysctl.html

**Reply by crocket** [reply]
```
# getcap /usr/bin/dnscrypt-proxy
/usr/bin/dnscrypt-proxy cap_net_bind_service=ep
```
/etc/firejail/dnscrypt-proxy.profile contains
```
caps.keep ipc_lock,net_bind_service,setgid,setuid,sys_chroot
```
But, if dnscrypt-proxy was executed as a non-root user in vpn network namespace and net.ipv4.ip_unprivileged_port_start=1024 in vpn network namespace, dnscrypt-proxy doesn't have permission to bind to port 53. Why? Is caps.keep broken?

**Reply by rusty-snake** [reply]
`man 2 prctl`:

> **PR_SET_NO_NEW_PRIVS** (since Linux 3.5)
              Set the calling thread's no_new_privs attribute to the
              value in arg2.  With no_new_privs set to 1, [execve(2)](https://man7.org/linux/man-pages/man2/execve.2.html)
              promises not to grant privileges to do anything that could
              not have been done without the [execve(2)](https://man7.org/linux/man-pages/man2/execve.2.html) call (for
              example, rendering the set-user-ID and set-group-ID mode
              bits, and file capabilities non-functional).  Once set,
              the no_new_privs attribute cannot be unset.  The setting
              of this attribute is inherited by children created by
              [fork(2)](https://man7.org/linux/man-pages/man2/fork.2.html) and [clone(2)](https://man7.org/linux/man-pages/man2/clone.2.html), and preserved across [execve(2)](https://man7.org/linux/man-pages/man2/execve.2.html).

**Reply by rusty-snake** [reply]
> Is caps.keep broken?

Nope

```console
$ firejail --quiet --noprofile --caps.keep=ipc_lock,net_bind_service,setgid,setuid,sys_chroot grep CapBnd /proc/self/status 
CapBnd:	00000000000444c
```

**Reply by crocket** [reply]
I think `man firejail` should mention file capabilities are ignored by `--nonewprivs`.

**Reply by rusty-snake** [reply]
> **--nonewprivs**
> Sets the NO_NEW_PRIVS prctl.  This ensures that child processes cannot acquire new privileges using execve(2);  in particular,  this  means that calling a suid binary (<h3>or one with file capabilities</h3>) does not result in an increase of privilege. This option is enabled by default if seccomp filter is activated.

**Reply by crocket** [reply]
Hell, it seems my brain was too tired to notice that.

**Comment by rusty-snake** [comment, marked as answer]
Summary

1. If you want to bind to a privileged port (`port < net.ipv4.ip_unprivileged_port_start`), you need to be root or have `CAP_NET_BIND_SERVICE`.
2. If you set `PR_SET_NO_NEW_PRIVS`, file capabilities are ignored.

**Reply by crocket** [reply]
I didn't know about `setpriv`. Perhaps, it should be mentioned in `man firejail` or somewhere else?

**Reply by rusty-snake** [reply]
Amb-caps don't seem to go through firejail. But inh-caps do, maybe you can move them to the eff-caps from outside by attaching? IDK.

(Assuming you want to use seccomp, ...), the best way is likely to start firejail as root and drop to a different user inside the sandbox.

_Or just use systemd … :speak_no_evil:_

**Reply by crocket** [reply]
I don't use systemd. I ended up resorting to `sysctl net.ipv4.ip_unprivileged_port_start=0`.

It seems apparmor can be better than firejail for system daemons.

**Comment by crocket** [comment]
@rusty-snake I updated https://github.com/netblue30/firejail/issues/4808 accordingly.

**Reply by kmk3** [reply]
Misc: Changed the title to match [the accepted answer](https://github.com/netblue30/firejail/discussions/4940#discussioncomment-2172869):

> How to make firejail more suitable for system daemons.

->

> How to launch dnsmasq as dnsmasq user in a network namespace?

@crocket Please be more specific when creating discussions to troubleshoot a
specific problem.
